src/backend/distributed/utils/metadata_cache.c

```diff
@@ -480,6 +492,60 @@ ResolveGroupShardPlacement(GroupShardPlacement *groupShardPlacement,
 
 
 /*
+ * LookupNodeForGroup searches the WorkerNodeHash for a worker which is a member of the
+ * given group and also readable (a primary if we're reading from primaries, a secondary
+ * if we're reading from secondaries). If such a node does not exist it emits an
+ * appropriate error message.
+ */
+static WorkerNode *
+LookupNodeForGroup(uint32 groupId)
+{
+	WorkerNode *workerNode = NULL;
+	HASH_SEQ_STATUS status;
+	HTAB *workerNodeHash = GetWorkerNodeHash();
+	bool foundAnyNodes = false;
+
+	hash_seq_init(&status, workerNodeHash);
+
+	while ((workerNode = hash_seq_search(&status)) != NULL)
+	{
+		uint32 workerNodeGroupId = workerNode->groupId;
+		if (workerNodeGroupId != groupId)
+		{
+			continue;
+		}
+
+		foundAnyNodes = true;
+
+		if (WorkerNodeIsReadable(workerNode))
+		{
+			hash_seq_term(&status);
+			return workerNode;
+		}
+	}
+
+	if (!foundAnyNodes)
+	{
+		ereport(ERROR, (errmsg("the metadata is inconsistent"),
+						errdetail("there is a placement in group %u but "
+								  "there are no nodes in that group", groupId)));
+	}
+
+	switch (ReadFromSecondaries)
+	{
+		case READ_FROM_SECONDARIES_NEVER:
+			ereport(ERROR, (errmsg("node group %u does not have a primary node",
+								   groupId)));
+		case READ_FROM_SECONDARIES_ALWAYS:
+			ereport(ERROR, (errmsg("node group %u does not have a secondary node",
+								   groupId)));
+		default:
+			ereport(FATAL, (errmsg("this should never happen")));
```

Comment:
maybe "unrecognised read_from_secondaries value" to help the person who will be adding a new option.